### PR TITLE
Improve rcl clock test coverage.

### PR DESCRIPTION
**Pull request by hidmic**
Precisely what the title says.

### CODE REVIEW

File: rcl/src/rcl/time.c [hunk 1]
```diff
@@ -366,6 +363,10 @@ rcl_set_ros_time_override(
   }
   rcl_time_jump_t time_jump;
   rcl_ros_clock_storage_t * storage = (rcl_ros_clock_storage_t *)clock->data;
+  if (!storage) {
```
Comment (brawner) [inline rcl/src/rcl/time.c:366]: Could this if statement just be a `RCL_CHECK_ARGUMENT_FOR_NULL` at the top of the function?

File: rcl/src/rcl/time.c [hunk 2]
```diff
@@ -453,19 +454,18 @@ rcl_clock_remove_jump_callback(
   }
 
   // Shrink size of the callback array
-  if (clock->num_jump_callbacks == 1) {
+  if (clock->num_jump_callbacks-- == 1) {
```
Comment (brawner) [inline rcl/src/rcl/time.c:457]: I think this change may need to go into a separate PR, based on the issue below.

File: rcl/src/rcl/time.c [hunk 3]
```diff
@@ -453,19 +454,18 @@ rcl_clock_remove_jump_callback(
...
     clock->allocator.deallocate(clock->jump_callbacks, clock->allocator.state);
     clock->jump_callbacks = NULL;
   } else {
     rcl_jump_callback_info_t * callbacks = clock->allocator.reallocate(
-      clock->jump_callbacks, sizeof(rcl_jump_callback_info_t) * (clock->num_jump_callbacks - 1),
+      clock->jump_callbacks, sizeof(rcl_jump_callback_info_t) * clock->num_jump_callbacks,
```
Comment (brawner) [inline rcl/src/rcl/time.c:462]: I'm having a hard time understanding what is happening here. At this point `clock->num_jump_callbacks` is 0. So realloc(ptr, 0) is happening. Why not just free the array? 

It looks like realloc of size 0 is now undefined behavior. I think this `if` statement could be cleaned up.

http://www.open-std.org/jtc1/sc22/wg14/www/docs/summary.htm#dr_400

File: rcl/test/rcl/test_time.cpp [hunk 4]
```diff
@@ -533,21 +499,15 @@ TEST(CLASSNAME(rcl_time, RMW_IMPLEMENTATION), rcl_time_fail_set_jump_callbacks)
...
-  auto * ros_clock =
-    static_cast<rcl_clock_t *>(allocator.allocate(sizeof(rcl_clock_t), allocator.state));
-  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
-  {
-    allocator.deallocate(ros_clock, allocator.state);
-  });
-  rcl_ret_t retval = rcl_ros_clock_init(ros_clock, &allocator);
-  ASSERT_EQ(retval, RCL_RET_OK) << rcl_get_error_string().str;
+  rcl_ret_t ret = rcl_ros_clock_init(&ros_clock, &allocator);
+  ASSERT_EQ(ret, RCL_RET_OK) << rcl_get_error_string().str;
```
Comment (brawner) [inline rcl/test/rcl/test_time.cpp:505]: ros_clock doesn't need to be cleaned up if ret is not OK?

File: rcl/test/rcl/test_time.cpp [hunk 5]
```diff
@@ -584,33 +544,27 @@ TEST(CLASSNAME(rcl_time, RMW_IMPLEMENTATION), rcl_time_forward_jump_callbacks) {
...
```
Comment (brawner) [inline rcl/test/rcl/test_time.cpp:563]: Same question.

File: rcl/test/rcl/test_time.cpp [hunk 6]
```diff
@@ -681,35 +630,30 @@ TEST(CLASSNAME(rcl_time, RMW_IMPLEMENTATION), rcl_clock_add_jump_callback) {
   rcl_jump_callback_t cb = reinterpret_cast<rcl_jump_callback_t>(0xBEEF);
   void * user_data = reinterpret_cast<void *>(0xCAFE);
 
-  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, rcl_clock_add_jump_callback(clock, threshold, NULL, NULL));
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, rcl_clock_add_jump_callback(&clock, threshold, NULL, NULL));
```
Comment (brawner) [inline rcl/test/rcl/test_time.cpp:633]: To ensure that the invalid_argument refers to a missing jump_callback, can you set `user_data` appropriately here.

File: rcl/test/rcl/test_time.cpp [hunk 7]
```diff
@@ -722,61 +666,80 @@ TEST(CLASSNAME(rcl_time, RMW_IMPLEMENTATION), rcl_clock_remove_jump_callback) {
   void * user_data3 = reinterpret_cast<void *>(0xBEAD);
   void * user_data4 = reinterpret_cast<void *>(0xDEED);
 
-  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, rcl_clock_remove_jump_callback(clock, NULL, NULL));
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, rcl_clock_remove_jump_callback(&clock, NULL, NULL));
```
Comment (brawner) [inline rcl/test/rcl/test_time.cpp:669]: Set user_data here.

File: rcl/src/rcl/time.c (hunk 3 above)
Comment (brawner) [reply rcl/src/rcl/time.c:462]: By undefined, I guess it's implementation defined. So this might not be portable.

File: rcl/src/rcl/time.c (hunk 2 above)
Comment (hidmic) [reply rcl/src/rcl/time.c:457]: It could go on a separate PR, but it goes hand in hand with the test that prevents a regression.

File: rcl/src/rcl/time.c (hunk 3 above)
Comment (hidmic) [reply rcl/src/rcl/time.c:462]: > At this point `clock->num_jump_callbacks` is `0`

Hmm for `clock->num_jump_callbacks` to be `0` here, it would've had to be `1` when the condition was evaluated, in which case it would've executed the first branch. Note that we're using postfix decrement, not prefix. I can change the condition if it's confusing though.

File: rcl/test/rcl/test_time.cpp (hunk 4 above)
Comment (hidmic) [reply rcl/test/rcl/test_time.cpp:505]: It doesn't. And I _think_ we do strive for that to be the case for every `rcl` "object" (and rightly so).

File: rcl/test/rcl/test_time.cpp (hunk 5 above)
Comment (hidmic) [reply rcl/test/rcl/test_time.cpp:563]: Same as above.

File: rcl/src/rcl/time.c (hunk 3 above)
Comment (brawner) [reply rcl/src/rcl/time.c:462]: You're right, I was confused when looking at the diff of this logic. The if clause is executed if num_jump_callbacks is 1 when entering the loop. It will never be 0 because this function will have returned before then. The else clause is evaluated when it is strictly greater than 1. Thus reallocate will be called with a size that is never 0.

The only other difference that might be important between the diffs is that if reallocate fails, `num_jump_callbacks` no longer matches the size of `jump_callbacks`. Should num_jump_callbacks be reincremented in that case?

File: rcl/src/rcl/time.c (hunk 2 above)
Comment (brawner) [reply rcl/src/rcl/time.c:457]: Is this fixing a bug and should it be backported, or is it just some cleanup?

File: rcl/src/rcl/time.c (hunk 3 above)
Comment (hidmic) [reply rcl/src/rcl/time.c:462]: > Should `num_jump_callbacks` be reincremented in that case?

I thought about it, but the [preceding loop](https://github.com/ros2/rcl/blob/7723308e7c7f7515590bd709525fd840b4aed254/rcl/src/rcl/time.c#L445) probably removed that callback by then. 

Also, a failed reallocation means the function failed to shrink storage, not that the callback itself wasn't removed successfully. A future deallocation wouldn't be affected and a future reallocation would be a no-op.

File: rcl/src/rcl/time.c (hunk 1 above)
Comment (hidmic) [reply rcl/src/rcl/time.c]: Done in https://github.com/ros2/rcl/pull/685/commits/012c120bad0e56a8da83e53ec07d4b3a1c635b60.

File: rcl/src/rcl/time.c (hunk 2 above)
Comment (hidmic) [reply rcl/src/rcl/time.c]: Hmm, that is a good point. I think it could be backported, yeah. It adds safety checks and ensure `rcl_remove_jump_callback()` fails a reallocation safely i.e. that you can continue despite a failed a reallocation.

File: rcl/test/rcl/test_time.cpp (hunk 7 above)
Comment (hidmic) [reply rcl/test/rcl/test_time.cpp]: Good point. See https://github.com/ros2/rcl/pull/685/commits/f72b7e87dea794f10c749363b68733632fb084fd.

File: rcl/test/rcl/test_time.cpp (hunk 6 above)
Comment (hidmic) [reply rcl/test/rcl/test_time.cpp]: Good point. See https://github.com/ros2/rcl/pull/685/commits/f72b7e87dea794f10c749363b68733632fb084fd.

**Comment by hidmic** [comment]
CI up to `rcl`:
* Linux (w/ coverage) [Build Status](http://ci.ros2.org/job/ci_linux/11184/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6463/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9135/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11098/)

**Comment by hidmic** [comment]
> I agree that the time.c changes should be considered for backporting.

Alright then. I'll rebase-merge instead of squashing so we can cherry pick and backport that commit later on. Does that sound good @brawner @Blast545 ?

**Comment by brawner** [comment]
I'm not a veteran with this code base, so if others are ok with rebase merging then sure. But I would lean toward the side of just separating them into different PRs. I can't exactly tell, but the changes to time.c don's seem related to the test file. By squash merging in separate PRs, the commit links to the PR directly, which has the discussion and review, which may be helpful for anyone looking through the source code.

**Comment by hidmic** [comment]
Tests here would fail without [this change](https://github.com/ros2/rcl/pull/685/files#diff-019482352ce3358b201f3806b28d0377R175) and [these changes](https://github.com/ros2/rcl/pull/685/files#diff-019482352ce3358b201f3806b28d0377R447-R460). The rest are mere cleanup. 

I can nonetheless split them up and make this PR depend on that PR. Discussion already took place here, so that PR would ultimately refer to this PR. But I don't mind either way.

**Comment by hidmic** [comment]
A clean rebase, but I'll re-run the pickiest CI (Windows' :) ) just to be safe.

**Comment by hidmic** [comment]
Alright, going in. Thanks for the reviews guys!
